chain/sync_manager.go

```diff
@@ -52,108 +65,467 @@ type SyncManager interface {
+	sm.state[id] = ws
+	sm.mx.Unlock()
+
+	go sm.worker(ws)
+}
+
+func (sm *syncManager) worker(ws *workerState) {
+	log.Infof("worker %d syncing in %s", ws.id, ws.ts)
+
+	start := build.Clock.Now()
+
+	ctx := context.WithValue(sm.ctx, syncStateKey{}, ws.ss)
+	err := sm.doSync(ctx, ws.ts)
+
+	ws.dt = build.Clock.Since(start)
+	log.Infof("worker %d done; took %s", ws.id, ws.dt)
+	select {
+	case sm.statusq <- workerStatus{id: ws.id, err: err}:
+	case <-sm.ctx.Done():
+	}
+}
+
+// selects the initial sync target by examining known peer heads; only called once for the initial
+// sync.
+func (sm *syncManager) selectInitialSyncTarget() (*types.TipSet, error) {
+	var buckets syncBucketSet
+
+	var peerHeads []*types.TipSet
+	for _, ts := range sm.heads {
+		peerHeads = append(peerHeads, ts)
+	}
+	// clear the map, we don't use it any longer
+	sm.heads = nil
+
+	sort.Slice(peerHeads, func(i, j int) bool {
+		return peerHeads[i].Height() < peerHeads[j].Height()
+	})
+
+	for _, ts := range peerHeads {
+		buckets.Insert(ts)
+	}
+
+	if len(buckets.buckets) > 1 {
+		log.Warn("caution, multiple distinct chains seen during head selections")
+		// TODO: we *could* refuse to sync here without user intervention.
+		// For now, just select the best cluster
+	}
+
+	return buckets.Heaviest(), nil
+}
+
+// adds a tipset to the potential sync targets; returns true if there is a a tipset to work on.
+// this could be either a restart, eg because there is no currently scheduled sync work or a worker
+// failed or a potential fork.
+func (sm *syncManager) addSyncTarget(ts *types.TipSet) (*types.TipSet, bool, error) {
+	// Note: we don't need the state lock here to access the active worker states, as the only
+	//       competing threads that may access it do so through State() which is read only.
+
+	// if we have recently synced this or any heavier tipset we just ignore it; this can happen
+	// with an empty worker set after we just finished syncing to a target
+	if sm.recent.Synced(ts) {
+		return nil, false, nil
+	}
+
+	// if the worker set is empty, we have finished syncing and were waiting for the next tipset
+	// in this case, we just return the tipset as work to be done
+	if len(sm.state) == 0 {
+		return ts, true, nil
+	}
+
+	// check if it is related to any active sync; if so insert into the pending sync queue
+	for _, ws := range sm.state {
+		if ts.Equals(ws.ts) {
+			// ignore it, we are already syncing it
+			return nil, false, nil
+		}
+
+		if ts.Parents() == ws.ts.Key() {
+			// schedule for syncing next; it's an extension of an active sync
+			sm.pend.Insert(ts)
+			return nil, false, nil
+		}
+	}
+
+	// check to see if it is related to any pending sync; if so insert it into the pending sync queue
+	if sm.pend.RelatedToAny(ts) {
+		sm.pend.Insert(ts)
+		return nil, false, nil
+	}
+
+	// it's not related to any active or pending sync; this could be a fork in which case we
+	// start a new worker to sync it, if it is *heavier* than any active or pending set;
```

Comment:
This comment implies that after the `RelatedToAny` calls what's left here is a potential fork case but the fork could have been catch in `RelatedToAny` (see comment there).